larpandoracontent/LArTwoDReco/LArCosmicRay/ClusterAssociation.h

```diff
@@ -0,0 +1,349 @@
+     *  @param  pDownstreamCluster the address of the downstream cluster
+     */    
+    ClusterPairAssociation(const pandora::CartesianVector &upstreamMergePoint, const pandora::CartesianVector &upstreamMergeDirection,
+        const pandora::CartesianVector &downstreamMergePoint, const pandora::CartesianVector &downstreamMergeDirection, const pandora::Cluster *pUpstreamCluster,
+        const pandora::Cluster *pDownstreamCluster);
+
+    /**
+     *  @brief  Returns the address of the upstream cluster
+     *
+     *  @return  Cluster the address of the upstream cluster
+     */
+    const pandora::Cluster *GetUpstreamCluster() const;
+
+    /**
+     *  @brief  Returns the address of the downstream cluster
+     *
+     *  @return  Cluster the address of the downstream cluster
+     */
+    const pandora::Cluster *GetDownstreamCluster() const;    
+
+private:
+    const pandora::Cluster    *m_pUpstreamCluster;      ///< The address of the upstream cluster
+    const pandora::Cluster    *m_pDownstreamCluster;    ///< The address of the downstream cluster
+};
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+inline ClusterAssociation::ClusterAssociation() :
+    m_upstreamMergePoint(pandora::CartesianVector(0.f, 0.f, 0.f)),
+    m_upstreamMergeDirection(pandora::CartesianVector(0.f, 0.f, 0.f)),
+    m_downstreamMergePoint(pandora::CartesianVector(0.f, 0.f, 0.f)),
+    m_downstreamMergeDirection(pandora::CartesianVector(0.f, 0.f, 0.f)),
+    m_connectingLineDirection(pandora::CartesianVector(0.f, 0.f, 0.f))
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline ClusterAssociation::ClusterAssociation(const pandora::CartesianVector &upstreamMergePoint, const pandora::CartesianVector &upstreamMergeDirection,
+        const pandora::CartesianVector &downstreamMergePoint, const pandora::CartesianVector &downstreamMergeDirection) :
+    m_upstreamMergePoint(upstreamMergePoint),
+    m_upstreamMergeDirection(upstreamMergeDirection),
+    m_downstreamMergePoint(downstreamMergePoint),
+    m_downstreamMergeDirection(downstreamMergeDirection),
+    m_connectingLineDirection(0.f, 0.f, 0.f)
+{
+    const pandora::CartesianVector connectingLineDirection(m_downstreamMergePoint.GetX() - m_upstreamMergePoint.GetX(), 0.f, m_downstreamMergePoint.GetZ() - m_upstreamMergePoint.GetZ());
+    m_connectingLineDirection = connectingLineDirection.GetUnitVector();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline bool ClusterAssociation::operator==(const ClusterAssociation &clusterAssociation) const
+{
+    return (m_upstreamMergePoint == clusterAssociation.GetUpstreamMergePoint() &&  m_upstreamMergePoint == clusterAssociation.GetUpstreamMergePoint() &&
```

Comment:
Looks like two conditions are repeated. Did you mean to include conditions on m_upstreamMergeDirection and m_downstreamMergeDirection, too?